### PR TITLE
New main layout

### PR DESCRIPTION
**Pull request by tybug**
one or two TODO's left in the code to deal with but this branch is practically ready for merge.

Will edit in a good description for the pr here when I write the changelog.

### CODE REVIEW

**Comment by InvisibleSymbol** [comment]
Can we do something about the comments? I understand that it's important to document your code but there is just too much unnecessary stuff...

Some examples:

- https://github.com/circleguard/circleguard/blob/15cf156a76c148f8ebf4fcc19338f704152cae03/circleguard/gui.py#L399-L401 I mean seriously what else could "setAcceptDrops" mean???
- https://github.com/circleguard/circleguard/blob/15cf156a76c148f8ebf4fcc19338f704152cae03/circleguard/gui.py#L490-L495  This project isn't a tutorial on how to do "drag and drop operations"...
- https://github.com/circleguard/circleguard/blob/15cf156a76c148f8ebf4fcc19338f704152cae03/circleguard/gui.py#L503-L507 Ignoring the concept for now, why the long comment with an example?

**Comment by tybug** [comment]
That's a fair concern, though my tolerance for comments is apparently slightly higher than yours.

`self.setAcceptDrops(True)` - wanted to emphasize that only widgets that set this will receive drop events, since that was unexpected behavior to me. I thought subclassing `dragEnterEvent` was all you needed to accept a drop event.

`resources for drag and drop operations` - there is no single page I can link to that explains all the things you need to do to make a qdrag work. The qt tutorial is *functional*, but doesn't explain everything it does. Those links are the result of 5-10 minutes of searching and looking through other, more useless links. In the case that I want to understand exactly what is necessary to make qdrag work in the future, I'd like some easy links to work off of.

`relative position` - I added the example because I thought leaving it at `same relative position` was unclear. On second thought it was probably fine to leave it at that.

To be honest, this mess of comments is compensating for the lack of class organization/comments. If circleguard as a whole were structured better, I don't think these massive inline comments would be necessary, since responsibilities would be more obviously distributed.

**Comment by tybug** [comment]
hm, the steal check should be separated by a vertical bar to let people specify replays1 and replays2. Either have a new button to the left of the delete button which cycles between two subsections and a single section, or always have it be two sections. Either way we very much need to provide this functionality

File: circleguard/widgets.py [hunk 1]
```diff
@@ -649,6 +599,9 @@ class FolderChooser(QFrame):
 
     def __init__(self, title, path=str(Path.home()), folder_mode=True, multiple_files=False, file_ending="osu! Beatmapfile (*.osu)", display_path=True):
         super(FolderChooser, self).__init__()
+        self.highlighted = False # for show_required
```
Comment (InvisibleSymbol) [inline circleguard/widgets.py]: unnecessary comment?

File: circleguard/widgets.py [hunk 2]
```diff
@@ -694,20 +652,46 @@ def set_dir(self):
...
+    around the QFrame.
+
+    To use on a class, replace the inheritence of QFrame with BorderWidget.
+    Possibly works with multiple inheritence (using this as a mixin), but I'm
+    not sure about that.
+    """
+    def __init__(self):
+        super().__init__()
+        self.setFrameStyle(QFrame.Box)
+
```
Comment (InvisibleSymbol) [inline circleguard/widgets.py]: remove debug class?

File: circleguard/gui.py [hunk 3]
```diff
@@ -1096,10 +1400,26 @@ def switch_theme(dark, accent=QColor(71, 174, 247)):
...
+                          "}\n"
+                          "LoadableW {\n"
+                            "\tborder: 1.5px solid #272727\n"
+                          "}\n"
+                          "CheckW {\n"
+                            "\tborder: 1.5px solid #272727\n"
+                          "}\n"
+                          "DragWidget {\n"
+                            "\tborder: 1.5px solid #272727\n"
+                          "}\n")
```
Comment (InvisibleSymbol) [inline circleguard/gui.py]: fix indentation (also below)

File: circleguard/gui.py [hunk 4]
```diff
@@ -1060,15 +1333,46 @@ def __init__(self):
...
+        # if we can provide good support for them.
+        # self.correction_max_angle_display = SliderBoxSetting("Max angle display ", "Replays with a set of "
+        #         "three points making an angle less than this (*and* also satisfying correction_min_distance_display) will be printed "
+        #         "to the console.", "correction_max_angle_display", 360)
+        self.correction_min_distance = SliderBoxSetting("Min distance", "Replays with a set of three points "
+                "where either the distance from AB or BC is greater than this (*and* also satisfying correction_max_angle) "
+                "will be stored so you can view them, and printed to the console.", "correction_min_distance", 100)
+        # self.correction_min_distance_display = SliderBoxSetting("Min distance display", "Replays with a set of "
+        #         "three points where either the distance from AB or BC is greater than this (*and* also satisfying "
+        #         "correction_max_angle_display) will be printed to the console.", "correction_min_distance_display", 100)
```
Comment (InvisibleSymbol) [inline circleguard/gui.py]: Please clean this up. This can be added when we need it but there is no reason for it to be a comment for now

File: circleguard/gui.py [hunk 5]
```diff
@@ -1002,7 +1277,8 @@ def __init__(self):
         # we want widgets to fill from top down,
         # being vertically centered looks weird
         self.layout.setAlignment(Qt.AlignTop)
-        self.info_label = QLabel("After running Comparisons, this tab will fill up with results")
+        self.info_label = QLabel("After running checks, this tab will fill up "
+                                 "with replays that can be played back.")
```
Comment (InvisibleSymbol) [inline circleguard/gui.py]: why the line break? If you want to follow some pep rule or something please apply it to everything

File: circleguard/gui.py [hunk 6]
```diff
@@ -973,14 +1248,14 @@ def reload_theme(self):
         switch_theme(get_setting("dark_theme"))
 
     def _open_settings(self):
-        overwrite_config()  # generate file with latest changes
+        overwrite_config() # generate file with latest changes
```
Comment (InvisibleSymbol) [inline circleguard/gui.py]: I suggest that we should follow PEP 8 (two spaces before inline comment)

File: circleguard/gui.py [hunk 7]
```diff
@@ -377,62 +748,39 @@ def scroll_to_bottom(self):
...
     def add_circleguard_run(self):
-        current_tab = self.tabs.currentIndex()
-        run_type = MainTab.TAB_REGISTER[current_tab]["name"]
-
-        # special case; visualize runs shouldn't get added to queue
-        if run_type == "VISUALIZE":
-            self.visualize([replay.data for replay in self.visualize_tab.replays])
+        checks = self.checks
+        if not checks:
+            # no checks added
```
Comment (InvisibleSymbol) [inline circleguard/gui.py]: check already explains it, why the comment

File: circleguard/gui.py [hunk 8]
```diff
@@ -286,71 +313,345 @@ def scroll_to_bottom(self):
...
+        # 1=all the way to the right/down, 0=all the way to the left/up
+        x_ratio = event.pos().x() / self.width()
+        y_ratio = event.pos().y() / self.height()
+        self.drag = QDrag(self)
+        # https://stackoverflow.com/a/53538805/12164878
+        pixmap = DragWidget(self.name, self.loadable_id).grab()
+        # put cursor in the same relative position on the dragwidget as
+        # it clicked on the real Loadable widget.
+        self.drag.setHotSpot(QPoint(pixmap.width() * x_ratio, pixmap.height() * y_ratio))
+        self.drag.setPixmap(pixmap)
```
Comment (InvisibleSymbol) [inline circleguard/gui.py]: weird concept. It's fine but just putting the cursor in the bottom right would be fine

File: circleguard/gui.py [hunk 9]
```diff
@@ -286,71 +313,345 @@ def scroll_to_bottom(self):
...
+        self.setLayout(layout)
+
+
+class LoadableW(QFrame):
+    """
+    A widget representing a circleguard.Loadable, which can be dragged onto
+    a DetectWidget. Keeps track of how many LoadableWidgets have been created
+    as a static ID attribute.
+
+    W standing for widget.
```
Comment (InvisibleSymbol) [inline circleguard/gui.py]: ...
sounds unnecessary

File: circleguard/settings.py [hunk 10]
```diff
@@ -140,38 +157,78 @@ def on_setting_changed_from_gui(self, value):
...
+        "correction_min_distance": 8,
+        "correction_min_distance_display": 8
     },
     "Appearance": {
         "dark_theme": False,
         "visualizer_info": True,
-        "visualizer_bg": False
+        "visualizer_black_bg": False,
+        "required_style": "QLineEdit { border: 1px solid red }\n"
+                          "WidgetCombiner { border: 1px solid red }"
```
[the remark this answers is not in the document]
Comment (tybug) [reply circleguard/settings.py]: This isn't a setting because I think people will want to change it, but more so because I need to access it at various points in the code (not just a single function) and making it a global is pretty nasty, so I made it a setting

File: circleguard/utils.py [hunk 11]
```diff
@@ -54,47 +56,40 @@ def get_idle_setting_str():
...
+    mod = Mod.NM
+    for i in range(2, len(mod_string) + 1, 2):
+        single_mod_string = mod_string[i - 2: i]
+        # there better only be one Mod that has an acronym matching ours, but a comp + 0 index works too
+        matching_mods = [mod for mod in Mod.ORDER if mod.short_name() == single_mod_string]
+        if not matching_mods:
+            raise InvalidModException(f"Invalid mod string (no matching mod found for {single_mod_string})")
+        mod += matching_mods[0]
+    return mod
+
```
[the remark this answers is not in the document]
Comment (tybug) [reply circleguard/utils.py]: yeah good point, will update circlecore to allow `ModCombination` instantiation via int or str

File: circleguard/widgets.py (hunk 2 above)
Comment (tybug) [reply circleguard/widgets.py]: could be useful in the future? I've already used, then removed, a class very similar to this twice when debugging. I'll probably just end up forgetting about it and re-looking up how to set a box frame. And it's really one doing one line of work, so...will remove

File: circleguard/gui.py (hunk 8 above)
Comment (tybug) [reply circleguard/gui.py]: Meh, this behavior appears pretty much everywhere (dragging icons on the mac doc, or tabs on chrome). The only difference here is we're using a drag widget of a different size than the original.

File: circleguard/gui.py (hunk 9 above)
Comment (tybug) [reply circleguard/gui.py]: might be obvious to you and me right now, but I don't think it will be obvious if I come back to this code in three months. `W` isn't exactly an unambiguous acronym

File: circleguard/gui.py (hunk 6 above)
Comment (tybug) [reply circleguard/gui.py]: I am frankly a bit shocked to find that pep 8 recommends two spaces; I've never followed that myself and can't recall noticing other code that does. The reasoning seems to be to make the comments stand out, but with editors displaying comments in a softer font already, that seems like a nonissue. I would prefer to not have to type twice the spaces when making a comment

File: circleguard/gui.py (hunk 5 above)
Comment (tybug) [reply circleguard/gui.py]: It's to avoid going over 80 characters in a line.

Screen Shot 2020-01-27 at 9 32 55 PM

My linter doesn't yell at me for it, I just try and follow it where possible. I don't think it's good to follow it religiously (basically all of settings.py goes over, which is fine - that particular file needs a lot of horizontal space for text), and reformatting the entire codebase to follow by hand would be tedious. It should be done at some point, but I've been holding off.

File: circleguard/gui.py (hunk 3 above)
Comment (tybug) [reply circleguard/gui.py]: what's the issue here? I formatted it to be how it would appear in a .css file. More readable imo.

**Comment by tybug** [comment]
I already use pylint. Here are my relevant vscode settings:

```
    "python.linting.pylintArgs": [
        "--disable=all",
        "--enable=F,E,unreachable,duplicate-key,unnecessary-semicolon,global-variable-not-assigned,unused-variable,binary-op-exception,bad-format-string,bad-open-mode"
    ],
```

which is different from the standard because I wanted to get rid of annoying warnings. See https://stackoverflow.com/a/54488818/12164878, where I got most of this list from

**Comment by tybug** [comment]
The one thing I like about pycodestyle is pointing out two lines between classes, one between methods, etc. However there are too many good reasons to break this (see `def my_excepthook(exctype, value, tb):` and `def init(self, *args, **kwargs):` at the top of `gui.py`, which do not have enough newlines before or after) that I wouldn't want it enabled constantly. And there's no way to selectively disable those instances without adding a new comment to the code, which I am very against - I conceded to pylint for `disable=no-name-in-module` at the very top of `gui.py`, since that's an important error message to keep enabled, but didn't like pyqt because of something going wrong with the c++ bindings

File: circleguard/settings.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (InvisibleSymbol) [reply circleguard/settings.py]: we really need a better solution for this but I guess it's okay for now.

File: circleguard/gui.py (hunk 3 above)
Comment (InvisibleSymbol) [reply circleguard/gui.py]: other than the linter screaming at my face I would prefer using a multiline string.

those `\n` are horrible

Comment (tybug) [reply circleguard/gui.py]: ah that's much better than using `\n` and `\t`, kind of forgot about multilines

File: circleguard/gui.py (hunk 5 above)
Comment (InvisibleSymbol) [reply circleguard/gui.py]: >I just try and follow it where possible.

This... I... I understand what you mean but we can't just do however we want it. The outcome is code that is horrible to read. I would much prefer following the rule or not.

>  and reformatting the entire codebase to follow by hand would be tedious.

Every IDE can do this automatically, you just have to manually check it before commiting.

> It should be done at some point, but I've been holding off.

I would prefer doing it all at once over just randomly changing some lines

File: circleguard/gui.py (hunk 6 above)
Comment (InvisibleSymbol) [reply circleguard/gui.py]: Sure, but please change all comments.

Comment (tybug) [reply circleguard/gui.py]: forgot there were more; changed them now

File: circleguard/gui.py (hunk 9 above)
Comment (InvisibleSymbol) [reply circleguard/gui.py]: Class is called `LoadableW`, the documentation calls it `LoadableWidgets`. You should be able to understand that `LoadableWidgets` means that the class is a widget.

Comment (tybug) [reply circleguard/gui.py]: that's true, as written the bottom line is unnecessary. But "how many LoadableWidgets" is outdated and should be changed to "how many LoadableWs" anyway, in which case an explicit explanation for the `W` suffix is appropriate I think

Comment (InvisibleSymbol) [reply circleguard/gui.py]: Yea, if you change it to `LoadableW`s I agree with keeping the explanation.

File: circleguard/gui.py (hunk 5 above)
Comment (tybug) [reply circleguard/gui.py]: IDEs can do it automatically - that doesn't mean they can do it well. Sometimes splitting a comment at a certain word could cause confusion or ambiguity ("did the writer of the comment mean for the next line to be a continuation of the comment on the previous line, or a new sentence?"), which all needs to be dealt with manually. The effort to reviewing that automated formatting is nontrivial.

And we can in fact to it however we want - I don't think that code is horrible to read at all, even when some sections ignore the 80 character limit and some don't. Take cacher.py in circlecore, which (as of the time of writing) is about half and half for following the 80 character limit. I think it looks better than if none of the file followed the 80 character limit.

Comment (InvisibleSymbol) [reply circleguard/gui.py]: In `catcher.py` there is [a single line](https://github.com/circleguard/circlecore/blob/master/circleguard/cacher.py#L78) above 150 chars, which doesn't look any worse/better when you split it to a new line. Therefore I would set the line length limit to 150 chars. There isn't a single line that is being split that is shorter than 150 chars.

You just changed the limit from a 80 char limit to a 150 char limit. I don't see your point of "sometimes it looks better, sometimes it doesn't".

> I don't think that code is horrible to read at all

Sorry, let me rephrase. While reading the code, the thing I would have in my mind would be "Why is this line being split for absolutely no reason?"

**Comment by InvisibleSymbol** [comment]
To me the 2 empty lines indicate that the previous function (and next line/function) has nothing to do with the current function.

For example, I instantly know that these two functions are independent because they have 2 spaces between them.


And in this case I instantly know that these functions are somehow connected (a class).


This makes big classes way easier to read imo, ofc this is mostly a preference thing, I'm open to arguments against doing it.

> And there's no way to selectively disable those instances without adding a new comment to the code

Please do either or, not selectively. That pylint even lets you selectively disable it is horrible design imo. Theses are design rules. They are meant to be followed or not. If you only follow the rule sometimes you aren't following it.

I would love if we could create a config file for pycodestyle (or pylint), see https://pycodestyle.pycqa.org/en/latest/intro.html#configuration

File: circleguard/gui.py (hunk 5 above)
Comment (tybug) [reply circleguard/gui.py]: > You just changed the limit from a 80 char limit to a 150 char limit.

not what's happening at all! `cacher.py` should be rewritten to use 80 character line limits. I didn't mean to hold it up as an idea file, more of an example where half of the file follows style guidelines and the other doesn't, and looks better than if the entire file didn't follow the guidelines.

>  "sometimes it looks better, sometimes it doesn't".

if you're referring to why I think it's appropriate for `settings.py` to ignore this limit, that's because it has *so many long strings*. The vertical length of the file would be massive if we limited those all to 80 characters. It just doesn't make sense to apply the style guidelines there when there's a block of long strings. If there were only one long string, it would make sense to split it across multiple lines.

The bottom half of `settings.py` still follows (going a couple characters over in some places, but close enough) the 80 character guideline. I think the file as a whole looks better than if the bottom half ignored this guideline.

**Comment by tybug** [comment]
Yeah no I like having two lines between classes and one between class methods, I think we should apply that (almost) everywhere.

> Theses are design rules. They are meant to be followed or not.

very hard disagree. There are situations where it is entirely appropriate to ignore the guidelines. (they're guidelines, by the way, not rules. I don't want to argue semantics when this discussion should be more logical than that, but they are called guidelines). Take the instance I was talking about above - we have this code in `gui.py` - 


```python
threading_init = threading.Thread.__init__
def init(self, *args, **kwargs):
    threading_init(self, *args, **kwargs)
    run_original = self.run

    def run_with_except_hook(*args2, **kwargs2):
        try:
            run_original(*args2, **kwargs2)
        except Exception:
            sys.excepthook(*sys.exc_info())
    self.run = run_with_except_hook
threading.Thread.__init__ = init
```

where the top (`threading_init = threading.Thread.__init__`) and bottom (`threading.Thread.__init__ = init`) lines are not separated from the function by a single line. Pycodestyle would yell at you for doing this (2 lines before and after functions!), but doing so would say "these lines that talk about threading_init are not related to the function below", when they absolutely are. The lines are intentionally bunched close to the function.

File: circleguard/gui.py (hunk 5 above)
Comment (InvisibleSymbol) [reply circleguard/gui.py]: > more of an example where half of the file follows style guidelines and the other doesn't

But that isn't the case! You didn't split a single line that is under 150 chars long, even though they are over 80 chars. You obviously followed the 150 char limit rule. You did not follow the 80 char rule sometimes. The freaking lines you split are still longer than 80 chars 

> if you're referring to why I think it's appropriate for settings.py to ignore this limit

I fully agree with ignoring the limit for settings.py, the long lines are data (In my eyes at least), not code.

> The bottom half of settings.py still follows (going a couple characters over in some places, but close enough) the 80 character guideline. I think the file as a whole looks better than if the bottom half ignored this guideline.

Again. There is not a single line that is being split in the bottom section. All code is shorter than 80 chars besides one comment (which can be ignore imo, since it isn't code). If you disabled the 80 char limit, and format the file so every line is as long as possible, _`settings.py` will not change by a single bit._

Comment (tybug) [reply circleguard/gui.py]: > You didn't split a single line that is under 150 chars long,

I split comments; I consider those just as valid as code for following 80 characters. You're right that I didn't split any line of code that was over 80 lines, but we're getting off track from the original debate. I'm not trying to say that `cacher.py` has been intentionally modified to fit within 80 characters, because in this case it hasn't. All I'm saying is that files look better when they *sometimes* follow the 80 character limit than when they don't at all.

Comment (InvisibleSymbol) [reply circleguard/gui.py]: In my eyes code and comments are 2 different things, but as you mentioned this is getting off track. We will keep it how it is.

**Comment by tybug** [comment]
what special kinds of loadables/checks do we still need to provide? The only one I can think of is a `User Check` which takes a user id and two spans, `span1` `span2`, doing the equivalent of `Map(span=span2) + ReplayMap(user_id)` for each top play in `span1`, with potentially an additional checkbox to do remod checks for the maps in `span1` as well.

**Comment by InvisibleSymbol** [comment]
Please change the names of the loadables. I might be able to do understand what you mean with `ReplayMap` but it isn't clear at first glance; I first misunderstood it as "Replays from a Map", tho it actually means "A Users replay on a Map". Then we have `MapUser` which it first thought is the same thing as `ReplayMap`, but is instead a mixture of `ReplayMap` and `Map`.

The gui is currently unusable if you try to run a simple steal check of the top 50. It took around 3 minutes for all comparisons to show up, the gui was unusable during this period.


The span from `Map` should have a sane default entry, maybe `1-50`

The id tag on a `MapUser` loadable is closer to the text than the other loadables 

> what special kinds of loadables/checks do we still need to provide? The only one I can think of is a User Check which takes a user id and two spans, span1 span2, doing the equivalent of Map(span=span2) + ReplayMap(user_id) for each top play in span1, with potentially an additional checkbox to do remod checks for the maps in span1 as well.

Isn't the `User` loadable exactly supposed to do that?

**Comment by tybug** [comment]
1) Name changes can come when we finalize the rest, I still need to think about good ones anyway
2) will investigate, this possibly only occurs on windows and not mac like previous frozen gui issues we had
4) well in circlecore a User represents a span of its top plays, not anything to do with the other replays on the maps. I think it would be best to keep that and add this separate user/remod check.

**Comment by tybug** [comment]
frozen gui is apparently just an issue with windows, 50 replays works fine for me. I forget what we did to fix it the last time this happened.

Having a default is a good idea. Only question is `1-50` or `1-100`, but `1-50` is probably better.

**Comment by tybug** [comment]
I can't reproduce the gui freezes so can't fix them. If you could try to see what's causing the issue that would be appreciated. The MapUser text for me is also still two tabs away, must be a cross OS issue again.

**Comment by tybug** [comment]
can't reproduce the gui freezing on my windows 8 at home either, either when loading 50 replays or comparing those 50 replays for replay stealing. I do see the MapUser issue though

File: circleguard/gui.py [hunk 12]
```diff
@@ -279,71 +305,416 @@ def __init__(self):
...
+        self.setAcceptDrops(True)
+
+        self.layout = QVBoxLayout()
+        self.layout.setAlignment(Qt.AlignTop)
+        self.setLayout(self.layout)
+
+    def dragEnterEvent(self, event):
+        # need to accept the event so qt gives us the DropEvent
+        # https://doc.qt.io/qt-5/qdropevent.html#details
+        event.acceptProposedAction()
```
Comment (InvisibleSymbol) [inline circleguard/gui.py]: Should probably check that there is `circleguard/` data in the mimedata; atm you can drop anything in a Check and it will try to parse that.

Comment (tybug) [reply circleguard/gui.py]: Good point.

took the liberty of rewriting our handling of mime type as well. `application/` type is for data handled by applications and `x` is for "non-standard" mime types. Don't think we were using the MIME standard correctly before.

File: circleguard/gui.py [hunk 13]
```diff
@@ -279,71 +307,421 @@ def __init__(self):
...
+    print_results_signal = pyqtSignal() # called after a run finishes to flush the results queue before printing "Done"
 
-    TAB_REGISTER = [
-        {"name": "MAP"},
-        {"name": "SCREEN"},
-        {"name": "LOCAL"},
-        {"name": "VERIFY"},
-        {"name": "VISUALIZE"}
-    ]
+    LOADABLES_COMBOBOX_REGISTRY = ["Replay Map", "Replay Path", "Map", "User", "Map User"]
```
Comment (InvisibleSymbol) [inline circleguard/gui.py]:
```suggestion
    LOADABLES_COMBOBOX_REGISTRY = ["User Replay on Map", "Local Replay", "Map", "User", "All Users Replays on Map"]
```
Last one is a bit meh but couldn't think of a better name

File: circleguard/gui.py [hunk 14]
```diff
@@ -279,71 +307,421 @@ def __init__(self):
...
+    CHECKS_COMBOBOX_REGISTRY = ["Replay Stealing", "Relax", "Aim Correction"]
```
Comment (InvisibleSymbol) [inline circleguard/gui.py]:
```suggestion
    CHECKS_COMBOBOX_REGISTRY = ["Replay Stealing and Remodding", "Relax", "Aim Correction"]
```

File: circleguard/gui.py [hunk 15]
```diff
@@ -353,12 +731,72 @@ def __init__(self):
...
+        layout.addWidget(self.terminal, 5, 0, 2, 16)
+        layout.addWidget(self.run_button, 7, 0, 1, 16)
+
         self.setLayout(layout)
-        self.switch_run_button()  # disable run button if there is no api key
+        self.check_run_button() # disable run button if there is no api key
+
+    # am well aware that there's much duplicated code between remove_loadable,
+    # remove_check, add_loadable, and add_check. Don't feel like writing
+    # more generic functions for them right now.
```
Comment (InvisibleSymbol) [inline circleguard/gui.py]: unnecessary comment

File: circleguard/gui.py [hunk 16]
```diff
@@ -613,26 +1052,43 @@ def print_results(self):
         try:
             while True:
                 result = self.q.get_nowait()
-                # self.visualize(result.replay1, result.replay2)
+                # self.visualize(result.replay1, result.replay2)]
```
Comment (InvisibleSymbol) [inline circleguard/gui.py]: delete

File: circleguard/utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (InvisibleSymbol) [reply circleguard/utils.py]: this is still here even tho it has been added to circlecore

File: circleguard/version.py [hunk 17]
```diff
@@ -1 +1 @@
-__version__ = "1.2.2"
+__version__ = "1.3.0.dev"
```
Comment (InvisibleSymbol) [inline circleguard/version.py]:
```suggestion
__version__ = "2.0.0.dev"
```
?

File: circleguard/visualizer.py [hunk 18]
```diff
@@ -671,7 +668,7 @@ def play_reverse(self):
         self._update_speed()
 
     def _update_speed(self):
-        self.renderer.clock.change_speed(float(self.speed_label.text()) * self.renderer.play_direction)
+        self.renderer.clock.change_speed(float(self.speed_label.text())*self.renderer.play_direction)
```
Comment (InvisibleSymbol) [inline circleguard/visualizer.py]: and same through the whole file

File: circleguard/gui.py (hunk 15 above)
Comment (tybug) [reply circleguard/gui.py]: disagree, it's a reminder to clean up the code later

File: circleguard/visualizer.py (hunk 18 above)
Comment (tybug) [reply circleguard/visualizer.py]: might have missed some but there's my pass

File: circleguard/widgets.py [hunk 19]
```diff
@@ -325,138 +334,372 @@ def __init__(self, tooltip):
...
+        self.layout.addWidget(self.user_id_input, 1, 0, 1, 8)
+        self.layout.addWidget(self.span_input, 2, 0, 1, 8)
+        self.layout.addWidget(self.mods_input, 3, 0, 1, 8)
+
+
+class MapUserW(LoadableW):
+    def __init__(self):
+        self.map_id_input = InputWidget("Map id", "", "id")
+        self.user_id_input = InputWidget("User id", "", "id")
+        self.span_input = InputWidget("Span", "", "normal")
```
Comment (InvisibleSymbol) [inline circleguard/widgets.py]: I suggest a default of 1-3. (or make it optional)

Comment (tybug) [reply circleguard/widgets.py]: if it were optional it would load all of them? If so I think it would be better to be explicit about it and fill it with an `"all"` placeholder text, since we don't have optional spans anywhere else, I think that would confuse people

Comment (InvisibleSymbol) [reply circleguard/widgets.py]: ye "all" is fine imo
